### ISSUE TITLE
Barcode multimarker example not working

### ISSUE THREAD

**Issue by enguerranws** [artoolkit/jsartoolkit5#34]
On the example Barcode multimarker: 
https://artoolkit.github.io/jsartoolkit5/examples/multimarker_barcode_threejs.html

I'm encountering this error:
```
Error opening pattern file '/00			' for reading.
artoolkit.min.js:1 Error processing multimarker config file '/multi_marker_0': Unable to load pattern '/00			'.
artoolkit.min.js:1 config data load error !!
artoolkit.min.js:1 ARToolKitJS(): Unable to set up AR multimarker.
```

The relevant file is https://github.com/artoolkit/jsartoolkit5/blob/master/examples/Data/multi-barcode-4x3.dat

On Chrome 52, MacOs.

**Comment by ThorstenBux** [comment]
@Neuroforge Can you have a look at that?

**Comment by Neuroforge** [comment]
Can you show me the folder structure that you have relating the page to the data file?

**Comment by enguerranws** [comment]
I'm having this trouble on the example online (https://artoolkit.github.io/jsartoolkit5/examples/multimarker_barcode_threejs.html), si I guess the folder structure is the same as the repo one?

**Comment by yuccai** [comment]
I think that they are space/tab characters just after the name of the marker. You should edit the dat file and remove them. Here is mine (renamed in txt because Github don't support dat format)
[multi-barcode-4x3.txt](https://github.com/artoolkit/jsartoolkit5/files/505965/multi-barcode-4x3.txt).
Please, state me if you succeed to run the example because I have a wrong transform matrix with Nan values in it.

**Comment by enguerranws** [comment]
@yuccai That did the trick, thanks. You should do a PR on that.

**Comment by yuccai** [comment]
@enguerranws  So you're example is now running correctly ? Don't you have Nan value in your transform matrix ?

**Comment by enguerranws** [comment]
I don't know if it's running correctly, however, I don't get the following error no more : 
`Error opening pattern file '/00         ' for reading.
artoolkit.min.js:1 Error processing multimarker config file '/multi_marker_0': Unable to load pattern '/00          '.
artoolkit.min.js:1 config data load error !!
artoolkit.min.js:1 ARToolKitJS(): Unable to set up AR multimarker.`

**Comment by MartinSoto** [comment]
I can confirm what @yuccai noticed: by removing the tab characters from the definition file, it is possible to read the multi-marker. The example code doesn't work though, and it always returns a broken transformation matrix containing a NaN value and otherwise zero or almost zero values.

Any help would be greatly appreciated, I'm working on a project right now that would benefit a lot from robust multi-marker recognition.

**Comment by MartinSoto** [comment]
To add to the previous comment: [this version](https://github.com/artoolkit/jsartoolkit5/commit/686d6159fd83c663e2ae541e17b25bd77038c189) seems to be the one that broke multimarker recognition. Reverting to the previous version makes it work for me.

**Comment by Sirormy** [comment]
Hi,
When I use multimarker, match marker trigger the event "getMultimarker"

```code
this.addEventListener('getMultiMarker', function(ev) {
  var obj = this.threeMultiMarkers[ev.data.multiMarkerId];
  if (obj) {
    obj.matrix.elements.set(ev.data.matrix);
    obj.visible = true;
  }
});
```

the ev.data.matrix value is `[0, 0, NaN, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]`
what`s the reason caused this ?

On MacOS Chrom 55

**Comment by DavideCapitola** [comment]
Hi,
I have tried to implement your code, but I still don’t get any relevant change.

This time I don’t get anymore the error as reported on the issue <https://github.com/artoolkit/jsartoolkit5/issues/43> in Github,
but the camera simply starts and shuts off after 3 seconds, and the video canvas dimension is set to 2px by 2px.


You can find a zip file with my code attached to this mail.
Do you have an idea of what can be the issue then?

Thanks in advance for your reply!


Met vriendelijke groet / Kind Regards,
 
Davide Zanon
Front-End Developer



Keizersgracht 205
1016 DS  Amsterdam

Office     + 31 20 845 2965 <tel:%2B%2031%2020%20845%202965>      
Mobile    + 31 6 39 664 432 <tel:%2B%2031%2020%20845%202965>
Web        www.capitola.nl <http://www.capitola.nl/>

**Comment by peterjwest** [comment]
Can someone fix this please?

The replacement `.dat` file stops it from crashing the tab outright, but I'm still not getting markers overlayed on the video.

**Comment by Sirormy** [comment]
the `multi-barcode-4x3.dat` has useless space, remove it, and the C source code has some bugs, fix it and use emscripten build js, the problem will be solved.

**Comment by jeromeetienne** [comment]
@Sirormy can you elaborate the bugs you found in the C source code ?

**Comment by Sirormy** [comment]
in [this file]( https://github.com/artoolkit/artoolkit5/blob/master/lib/SRC/ARMulti/arMultiReadConfigFile.c) line: 197

```c
marker_info->prevF      = 0;
```

after this line, add

```c
marker_info->min_submarker = num/3;
```

it workd on me.

without this line, matrix is bad.

And also, you need use emscripten to rebuild the whole project, the new artoolkit.js will work.

**Comment by Wavewash** [comment]
In trying to follow along with @Sirormy and the build instructions:
https://artoolkit.github.io/jsartoolkit5/

What are the "Install build tools" - are these just the node build tools?

Step 7 states:
Make sure EMSCRIPTEN env variable is set (e.g. EMSCRIPTEN=/usr/lib/emsdk_portable/emscripten/master/ node tools/makem.js

The line doesn't have a closing parenthesis and the Path environment variable doesn't seem to be valid with node tools/makem.js at the end. Was this the intent?

Any help would be appreciated.

**Comment by Flikker** [comment]
Hi guys! Do you reckon it'll be possible to fix the multimarkers? I've tried making the changes and rebuilding, but been running into problems being on a PC! 
As well as the lack of matrix data, it also doesn't seem to allow 3d models to be placed on a multimarker. I checked with an artoolkit.min.js from an earlier commit that had the matrix data (camera was too slow to use this version in my actual project) and 3d models still didn't appear on top of it, so I think the two issues are separate.
Also, just to say thanks for everyone's work on this- I've really enjoyed using it and it's a fantastic project.

**Comment by dekwilde** [comment]
@Sirormy can you publish this rebuild artoolkit.js that you talk about in this reply: 
"
in this file line: 197

marker_info->prevF      = 0;
after this line, add

marker_info->min_submarker = num/3;
it workd on me.

without this line, matrix is bad.

And also, you need use emscripten to rebuild the whole project, the new artoolkit.js will work.
"
thank you...

**Comment by Sirormy** [comment]
@dekwilde 
Hi,
this is the built file
[artoolkit.min.js](https://drive.google.com/open?id=0B_HAsoqFl10AY2FrU2gweENWRzA)

**Comment by dekwilde** [comment]
@Sirormy awesome now it's working... thank you!!!

**Comment by bwilytsch** [comment]
@Sirormy has this update been pushed to npm as well? This would be amazing!

**Comment by Sirormy** [comment]
@bwilytsch no, it`s just built  by myself, maybe I should give a pull request?

**Comment by bwilytsch** [comment]
@Sirormy I would highly appreciate it or your unminified artoolkit.js would already be super helpful! Thanks a lot!

**Comment by bwilytsch** [comment]
Does anyone have a working example for barcode multimarkers? Everything online is broken.

**Comment by Sirormy** [comment]
@bwilytsch  Sorry, I`m not sure how to use emscripten to build a unminified version, I just take a look at docs when I use it. May be this file is useful for you,  [artoolkit.debug.js](https://drive.google.com/file/d/0B_HAsoqFl10AWnR4ai12OG1SQm8/view?usp=sharing), it`s a debug js built in the maintime with artoolkit.min.js, it`s part of source? And [ARBindEM](https://github.com/artoolkit/jsartoolkit5/blob/master/emscripten/ARBindEM.cpp) is the bridge between js and c implementation, and [ARToolKitJS](https://github.com/artoolkit/jsartoolkit5/blob/master/emscripten/ARToolKitJS.cpp) is the c implmentation, I think these are enough for developing.

**Comment by nylki** [comment]
The examples `multimarker_barcode_threejs.html` and `multimarker_pattern_threejs.html` are still not working on Chrome or Firefox. The barcode example crashes/freezes the video once aruco markers are in the video.
The marker example doesn't crash but also doesn't display anything when the printed *A B C D G F* is in the video feed.

As some of you have apparently made it work with a newer emscripten build of artoolkit5, it would be great to see those examples working on the website. :)

I'll try to make a new emscripten build, to check wheter the examples are working.

**Comment by nylki** [comment]
Quick update: I succeded in compiling a fresh jsartoolkit5 via emscripten after some changes on Mac OS (using different version of `llvm-ar` because of errors and removing the `NATIVE_OPTIMIZER` variable from ~/emscripten config file.

However, the same issues as described above exist in those two examples. Any idea why? /cc @Sirormy @ThorstenBux

**Comment by Sirormy** [comment]
@nylki try to check your `Data/multi/xxx.dat` file, remove useless tab or space, maybe it will work.

**Comment by mflux** [comment]
This example still needs to be fixed. From latest ARToolkit build I ran the multi target example and still getting this error.

Only way to fix it was to replace multi-barcode-4x3.dat with the content posted by @yuccai 

https://github.com/artoolkit/jsartoolkit5/files/505965/multi-barcode-4x3.txt
